<?xml version="1.0"?>
<launch>
    <!-- MAVROS posix SITL environment launch script -->
    <!-- launches Gazebo environment and 2x: MAVROS, PX4 SITL, and spawns vehicle -->
    <!-- vehicle model and world -->
    <arg name="est" default="ekf2" />
    <!-- <arg name="world" default="$(find dmpc_uav_ad_hoc)/worlds/disaster.world" /> -->
    <arg name="world" default="$(find dmpc_uav_ad_hoc)/worlds/empty.world" />
    <!-- gazebo configs -->
    <arg name="gui" default="false" />
    <arg name="debug" default="false" />
    <arg name="verbose" default="false" />
    <arg name="paused" default="false" />
    <param name="use_sim_time" value="true" />
    <!-- Gazebo sim -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="gui" value="$(arg gui)" />
        <arg name="world_name" value="$(arg world)" />
        <arg name="debug" value="$(arg debug)" />
        <arg name="verbose" value="$(arg verbose)" />
        <arg name="paused" value="$(arg paused)" />
    </include>
    <param name="robot_description" textfile="$(find dmpc_uav_ad_hoc)/rviz/iris.urdf" />

    <group ns="iris_0" if="$(eval arg('uav_num') >= 1)">
        <remap from="/iris_0/mavros/vision_pose/pose" to="/iris_0/real_pose" />
        <arg name="ID" value="0" />
        <arg name="ID_in_group" value="0" />
        <arg name="fcu_url" default="udp://:24540@localhost:34580" />
        <include file="$(find px4)/launch/single_vehicle_spawn_xtd.launch">
            <arg name="x" value="0" />
            <arg name="y" value="3" />
            <arg name="z" value="0.5" />
            <arg name="R" value="0" />
            <arg name="P" value="0" />
            <arg name="Y" value="0" />
            <arg name="vehicle" value="iris" />
            <arg name="sdf" value="iris_stereo_camera" />
            <arg name="mavlink_udp_port" value="18570" />
            <arg name="mavlink_tcp_port" value="4560" />
            <arg name="ID" value="$(arg ID)" />
            <arg name="ID_in_group" value="$(arg ID_in_group)" />
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="" />
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))" />
            <arg name="tgt_component" value="1" />
        </include>
    </group>
    <group ns="iris_1" if="$(eval arg('uav_num') >= 2)">
        <remap from="/iris_1/mavros/vision_pose/pose" to="/iris_1/real_pose" />
        <arg name="ID" value="1" />
        <arg name="ID_in_group" value="1" />
        <arg name="fcu_url" default="udp://:24541@localhost:34581" />
        <include file="$(find px4)/launch/single_vehicle_spawn_xtd.launch">
            <arg name="x" value="3" />
            <arg name="y" value="3" />
            <arg name="z" value="0.5" />
            <arg name="R" value="0" />
            <arg name="P" value="0" />
            <arg name="Y" value="0" />
            <arg name="vehicle" value="iris" />
            <arg name="sdf" value="iris_stereo_camera" />
            <arg name="mavlink_udp_port" value="18571" />
            <arg name="mavlink_tcp_port" value="4561" />
            <arg name="ID" value="$(arg ID)" />
            <arg name="ID_in_group" value="$(arg ID_in_group)" />
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="" />
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))" />
            <arg name="tgt_component" value="1" />
        </include>
    </group>
    <group ns="iris_2" if="$(eval arg('uav_num') >= 3)">
        <remap from="/iris_2/mavros/vision_pose/pose" to="/iris_2/real_pose" />
        <arg name="ID" value="2" />
        <arg name="ID_in_group" value="2" />
        <arg name="fcu_url" default="udp://:24542@localhost:34582" />
        <include file="$(find px4)/launch/single_vehicle_spawn_xtd.launch">
            <arg name="x" value="6" />
            <arg name="y" value="3" />
            <arg name="z" value="0.5" />
            <arg name="R" value="0" />
            <arg name="P" value="0" />
            <arg name="Y" value="0" />
            <arg name="vehicle" value="iris" />
            <arg name="sdf" value="iris_stereo_camera" />
            <arg name="mavlink_udp_port" value="18572" />
            <arg name="mavlink_tcp_port" value="4562" />
            <arg name="ID" value="$(arg ID)" />
            <arg name="ID_in_group" value="$(arg ID_in_group)" />
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="" />
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))" />
            <arg name="tgt_component" value="1" />
        </include>
    </group>

    <group ns="iris_3" if="$(eval arg('uav_num') >= 4)">
        <remap from="/iris_3/mavros/vision_pose/pose" to="/iris_3/real_pose" />
        <arg name="ID" value="3" />
        <arg name="ID_in_group" value="3" />
        <arg name="fcu_url" default="udp://:24543@localhost:34583" />
        <include file="$(find px4)/launch/single_vehicle_spawn_xtd.launch">
            <arg name="x" value="0" />
            <arg name="y" value="6" />
            <arg name="z" value="0.5" />
            <arg name="R" value="0" />
            <arg name="P" value="0" />
            <arg name="Y" value="0" />
            <arg name="vehicle" value="iris" />
            <arg name="sdf" value="iris_stereo_camera" />
            <arg name="mavlink_udp_port" value="18573" />
            <arg name="mavlink_tcp_port" value="4563" />
            <arg name="ID" value="$(arg ID)" />
            <arg name="ID_in_group" value="$(arg ID_in_group)" />
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="" />
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))" />
            <arg name="tgt_component" value="1" />
        </include>
    </group>
    <group ns="iris_4" if="$(eval arg('uav_num') >= 5)">
        <remap from="/iris_4/mavros/vision_pose/pose" to="/iris_4/real_pose" />
        <arg name="ID" value="4" />
        <arg name="ID_in_group" value="4" />
        <arg name="fcu_url" default="udp://:24544@localhost:34584" />
        <include file="$(find px4)/launch/single_vehicle_spawn_xtd.launch">
            <arg name="x" value="3" />
            <arg name="y" value="6" />
            <arg name="z" value="0.5" />
            <arg name="R" value="0" />
            <arg name="P" value="0" />
            <arg name="Y" value="0" />
            <arg name="vehicle" value="iris" />
            <arg name="sdf" value="iris_stereo_camera" />
            <arg name="mavlink_udp_port" value="18574" />
            <arg name="mavlink_tcp_port" value="4564" />
            <arg name="ID" value="$(arg ID)" />
            <arg name="ID_in_group" value="$(arg ID_in_group)" />
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="" />
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))" />
            <arg name="tgt_component" value="1" />
        </include>
    </group>
    <group ns="iris_5" if="$(eval arg('uav_num') >= 6)">
        <remap from="/iris_5/mavros/vision_pose/pose" to="/iris_5/real_pose" />
        <arg name="ID" value="5" />
        <arg name="ID_in_group" value="5" />
        <arg name="fcu_url" default="udp://:24545@localhost:34585" />
        <include file="$(find px4)/launch/single_vehicle_spawn_xtd.launch">
            <arg name="x" value="6" />
            <arg name="y" value="6" />
            <arg name="z" value="0.5" />
            <arg name="R" value="0" />
            <arg name="P" value="0" />
            <arg name="Y" value="0" />
            <arg name="vehicle" value="iris" />
            <arg name="sdf" value="iris_stereo_camera" />
            <arg name="mavlink_udp_port" value="18575" />
            <arg name="mavlink_tcp_port" value="4565" />
            <arg name="ID" value="$(arg ID)" />
            <arg name="ID_in_group" value="$(arg ID_in_group)" />
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="" />
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))" />
            <arg name="tgt_component" value="1" />
        </include>
    </group>
    <group ns="iris_6" if="$(eval arg('uav_num') >= 7)">
        <remap from="/iris_6/mavros/vision_pose/pose" to="/iris_6/real_pose" />
        <arg name="ID" value="6" />
        <arg name="ID_in_group" value="6" />
        <arg name="fcu_url" default="udp://:24546@localhost:34586" />
        <include file="$(find px4)/launch/single_vehicle_spawn_xtd.launch">
            <arg name="x" value="0" />
            <arg name="y" value="9" />
            <arg name="z" value="0.5" />
            <arg name="R" value="0" />
            <arg name="P" value="0" />
            <arg name="Y" value="0" />
            <arg name="vehicle" value="iris" />
            <arg name="sdf" value="iris_stereo_camera" />
            <arg name="mavlink_udp_port" value="18576" />
            <arg name="mavlink_tcp_port" value="4566" />
            <arg name="ID" value="$(arg ID)" />
            <arg name="ID_in_group" value="$(arg ID_in_group)" />
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="" />
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))" />
            <arg name="tgt_component" value="1" />
        </include>
    </group>
    <group ns="iris_7" if="$(eval arg('uav_num') >= 8)">
        <remap from="/iris_7/mavros/vision_pose/pose" to="/iris_7/real_pose" />
        <arg name="ID" value="7" />
        <arg name="ID_in_group" value="7" />
        <arg name="fcu_url" default="udp://:24547@localhost:34587" />
        <include file="$(find px4)/launch/single_vehicle_spawn_xtd.launch">
            <arg name="x" value="3" />
            <arg name="y" value="9" />
            <arg name="z" value="0.5" />
            <arg name="R" value="0" />
            <arg name="P" value="0" />
            <arg name="Y" value="0" />
            <arg name="vehicle" value="iris" />
            <arg name="sdf" value="iris_stereo_camera" />
            <arg name="mavlink_udp_port" value="18577" />
            <arg name="mavlink_tcp_port" value="4567" />
            <arg name="ID" value="$(arg ID)" />
            <arg name="ID_in_group" value="$(arg ID_in_group)" />
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="" />
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))" />
            <arg name="tgt_component" value="1" />
        </include>
    </group>
    <group ns="iris_8" if="$(eval arg('uav_num') >= 9)">
        <remap from="/iris_8/mavros/vision_pose/pose" to="/iris_8/real_pose" />
        <arg name="ID" value="8" />
        <arg name="ID_in_group" value="8" />
        <arg name="fcu_url" default="udp://:24548@localhost:34588" />
        <include file="$(find px4)/launch/single_vehicle_spawn_xtd.launch">
            <arg name="x" value="6" />
            <arg name="y" value="9" />
            <arg name="z" value="0.5" />
            <arg name="R" value="0" />
            <arg name="P" value="0" />
            <arg name="Y" value="0" />
            <arg name="vehicle" value="iris" />
            <arg name="sdf" value="iris_stereo_camera" />
            <arg name="mavlink_udp_port" value="18578" />
            <arg name="mavlink_tcp_port" value="4568" />
            <arg name="ID" value="$(arg ID)" />
            <arg name="ID_in_group" value="$(arg ID_in_group)" />
        </include>
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)" />
            <arg name="gcs_url" value="" />
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))" />
            <arg name="tgt_component" value="1" />
        </include>
    </group>

    <node pkg="dmpc_uav_ad_hoc" name="model_state_pub_node" type="model_state_pub_node"
        output="screen" launch-prefix="bash -c 'sleep 1.0; $0 $@';">
        <param name="uav_num" value="$(arg uav_num)" />
        <param name="pub_freq" value="60.0" />
        <param name="pub_tf" value="true" />
        <param name="state_source" value="topic" />
    </node>
</launch>